### PR TITLE
Implement MacOS FileWatcher

### LINKED ISSUE #11473: Implement FileWatcher for macOS
Currently the FileWatcher has an implementation for Windows and Linux but not macOS. This is required for plugin hot reloading which is very useful for plugin development.

Also required for #11439.

### PR DESCRIPTION
**Pull request by umar-ahmed**
Closes #11473

### CODE REVIEW

File: src/openrct2/core/FileScanner.h [hunk 1]
```diff
@@ -16,18 +16,21 @@
...
-    uint64_t Size;
-    uint64_t LastModified;
-};
+    struct FileInfo
+    {
+        const utf8* Name;
+        uint64_t Size;
+        uint64_t LastModified;
+    };
+} // namespace FileScanner
```
Comment (umar-ahmed) [inline src/openrct2/core/FileScanner.h:27]: `FileInfo` has a conflict with an internal struct in `CoreServices`

File: src/openrct2/entity/Guest.cpp [hunk 2]
```diff
@@ -1857,9 +1857,9 @@ Ride* Guest::FindBestRideToGoOn()
     return mostExcitingRide;
 }
 
-BitSet<OpenRCT2::Limits::MaxRidesInPark> Guest::FindRidesToGoOn()
+OpenRCT2::BitSet<OpenRCT2::Limits::MaxRidesInPark> Guest::FindRidesToGoOn()
```
Comment (umar-ahmed) [inline src/openrct2/entity/Guest.cpp:1860]: `BitSet` has a conflict with an internal struct in `CoreServices`

File: src/openrct2/core/FileWatcher.cpp [hunk 3]
```diff
@@ -82,6 +85,39 @@ FileWatcher::WatchDescriptor::~WatchDescriptor()
 }
 #endif
 
+#if defined(__APPLE__)
+
+static int eventModified = kFSEventStreamEventFlagItemFinderInfoMod | kFSEventStreamEventFlagItemModified
+    | kFSEventStreamEventFlagItemInodeMetaMod | kFSEventStreamEventFlagItemChangeOwner | kFSEventStreamEventFlagItemXattrMod;
+
+static int eventRenamed = kFSEventStreamEventFlagItemCreated | kFSEventStreamEventFlagItemRemoved
+    | kFSEventStreamEventFlagItemRenamed;
```
Comment (umar-ahmed) [inline src/openrct2/core/FileWatcher.cpp:94]: These can probably be reduced to just the events related to updates

File: cmake/platform.cmake [hunk 4]
```diff
@@ -3,7 +3,7 @@
 function(target_link_platform_libraries target)
 
 if (APPLE)
-    target_link_libraries(${target} "-framework Cocoa")
+    target_link_libraries(${target} "-framework Cocoa -framework CoreServices")
```
Comment (umar-ahmed) [inline cmake/platform.cmake:6]: fsevents requires CoreServices to be linked

Comment (umar-ahmed) [reply cmake/platform.cmake:6]: https://developer.apple.com/library/archive/documentation/Darwin/Conceptual/FSEvents_ProgGuide/UsingtheFSEventsFramework/UsingtheFSEventsFramework.html#//apple_ref/doc/uid/TP40005289-CH4-SW4

**Review by Gymnasiast — APPROVED** [review]
I haven’t been able to test this, but since all the changes are limited to macOS (except the renames) I don’t consider the risk of fallout very high. Though I would love if a macOS user could verify the workings.

**Comment by Gymnasiast** [comment]
Are you confident enough this will work? Since macOS currently does not have hot reload at all, I don’t mind taking a "gamble" here.

**Comment by umar-ahmed** [comment]
> Are you confident enough this will work? Since macOS currently does not have hot reload at all, I don’t mind taking a "gamble" here.

@Gymnasiast With the latest commit I have this working on my own M1 Mac. However, @ltsSmitty was having some issues with my latest commit. I'd like to take a bit more time to debug because it seems like the commit before that was working for him, and I'd like to get a better understanding of what regressed there.

**Comment by Gymnasiast** [comment]
No problem, I have marked your PR as draft. Once you’re satisfied it’s working, please mark it as ready for review.

**Comment by ltsSmitty** [comment]
I'll happily continue to review as needed.

**Comment by ltsSmitty** [comment]
Yep, this version works as expected! LGTM.

**Comment by umar-ahmed** [comment]
@Gymnasiast I think this should be good to go now. Feel free to re-review and merge

**Review by tupaschoal — COMMENTED** [review]
Should we add a changelog entry for this one? Seems like a useful enhancement for Mac players

**Comment by Gymnasiast** [comment]
I already did.

**Pull request merged**